Add support for tensorflow-nufft

chaithyagr opened the pull request:
This PR adds supports for tensorflow-nufft through the same interface.
We also bump up the tensorflow version to the latest

chaithyagr commented:
Speedup of nearly 7-8 x with this library...

zaccharieramzi reviewed (changes requested):
This looks great, thanks @chaithyagr !! Do you mind adding in a test for this implementation in `fastmri_recon/tests/models/subclassed_models/ncpdnet_test.py` ?

Not necessarily to test the output, but just to test that it runs

chaithyagr commented on fastmri_recon/models/utils/fourier.py:
```diff
@@ -1,9 +1,15 @@
 from tfkbnufft.kbnufft import KbNufftModule
-
+try:
+    import tensorflow_nufft as tfft
```
tfnufft ?

zaccharieramzi replied:
Yup perfect

chaithyagr commented:
The test may not work as it needs CUDA and a GPU if I am right... I can have one which can run locally when needed..

zaccharieramzi commented:
Yes even a local test is great!

zaccharieramzi commented:
And can you make sure to have all tests green ?

chaithyagr commented:
Lol yes on it. No idea what these errors are. Mostly it's due to the switch to tf 2.8

chaithyagr commented:
This mostly shall make it green. I had good tests locally. I will add some quick tests now for tfnufft checks

zaccharieramzi commented on .vscode/launch.json:
```diff
@@ -0,0 +1,25 @@
+{
```
Can you remove this?

chaithyagr replied:
Ahhh can we add `.vscode` in gitignore ? :P

zaccharieramzi replied:
ofc! please do

zaccharieramzi reviewed (changes requested):
Small leftover file

chaithyagr commented:
Removed, and added tensorflow-nufft. @zaccharieramzi can you recheck if you are happy with the new flow to test if we are on CI and then run / not run `tensorflow-nufft` implementation?

zaccharieramzi commented on fastmri_recon/tests/training_scripts/ncpdnet_train_test.py:
```diff
@@ -11,25 +10,26 @@
     nc_train.LOGS_DIR = create_full_fastmri_test_tmp_dataset['logs_tmp_dir']
     nc_train.CHECKPOINTS_DIR = create_full_fastmri_test_tmp_dataset['checkpoints_tmp_dir']
     nc_train.n_volumes_train_fastmri = 2
+    if nufft_implementation == 'tensorflow-nufft' and os.environ.get('CI', False) == True:
```
```suggestion
    if nufft_implementation == 'tensorflow-nufft' and os.environ.get('CI', False):
```

zaccharieramzi replied:
ah actually wait, are you doing this for a specific reason?

zaccharieramzi commented on fastmri_recon/tests/training_scripts/unet_nc_train_test.py:
```diff
@@ -7,7 +8,10 @@
-def test_train_unet_nc(create_full_fastmri_test_tmp_dataset, kwargs):
+@pytest.mark.parametrize('nufft_implementation',['tfkbnufft', 'tensorflow-nufft'])
+def test_train_unet_nc(create_full_fastmri_test_tmp_dataset, kwargs, nufft_implementation):
+    if nufft_implementation == 'tensorflow-nufft' and os.environ.get('CI', False) == True:
```
```suggestion
    if nufft_implementation == 'tensorflow-nufft' and os.environ.get('CI', False):
```

chaithyagr replied:
Ahhhh no, i got it wrong... It is:

`os.environ.get('CI', False) == "True"`

zaccharieramzi replied:
and why do you do this ?

chaithyagr replied:
Well environment variables are strings 
 So the Boolean value is stored as a string which is "True" or "False"

zaccharieramzi replied:
can't you make it a `bool` when creating it? 

FYI:
```
bool('True')
```
returns 
```
True
```

chaithyagr replied:
Also bool('False') returns True :P

zaccharieramzi commented on fastmri_recon/tests/conftest.py:
```diff
@@ -1,10 +1,10 @@
 import h5py
+import os
```
If you can put it at the top (as it is a built-in library), it would be awesome

zaccharieramzi reviewed (dismissed):
Ah actually one last thing I forgot (I remembered when seeing the training script): can you add the option of the nufft implem to the evaluation scripts as well?

Other than a few nitpicks, LGTM

chaithyagr commented:
The tests passed earlier, just have a gitignore update. @zaccharieramzi this can be merged and can you please make a new release as well?

chaithyagr commented:
Ahhh missed your comment, on it

zaccharieramzi reviewed (approved):
I think the `== 'True'` is a bit convoluted but I am fine with it.
LGTM
